// mithril@2.2.10/_performance.old/components/repeated-tree.js
class RepeatedHeader {
	view() {
		return m("header",
			m("h1.asdf", "a ", "b", " c ", 0, " d"),
			m("nav",
				m("a", {href: "/foo"}, "Foo"),
				m("a", {href: "/bar"}, "Bar")
			)
		)
	}
}

class RepeatedForm {
	view() {
		return m("form", {onsubmit() {}},
			m("input", {type: "checkbox", checked: true}),
			m("input", {type: "checkbox", checked: false}),
			m("fieldset",
				m("label",
					m("input", {type: "radio", checked: true})
				),
				m("label",
					m("input", {type: "radio"})
				)
			),
			m(RepeatedButtonBar, null)
		)
	}
}

class RepeatedButtonBar {
	view() {
		return m(".button-bar",
			m(RepeatedButton,
				{style: "width:10px; height:10px; border:1px solid #FFF;"},
				"Normal CSS"
			),
			m(RepeatedButton,
				{style: "top:0 ; right: 20"},
				"Poor CSS"
			),
			m(RepeatedButton,
				{style: "invalid-prop:1;padding:1px;font:12px/1.1 arial,sans-serif;", icon: true},
				"Poorer CSS"
			),
			m(RepeatedButton,
				{style: {margin: 0, padding: "10px", overflow: "visible"}},
				"Object CSS"
			)
		)
	}
}

class RepeatedButton {
	view({attrs}) {
		return m("button", m.censor(attrs))
	}
}

class RepeatedMain {
	view() {
		return m(RepeatedForm)
	}
}

class RepeatedRoot {
	view() {
		return m("div.foo.bar[data-foo=bar]",
			{p: 2},
			m(RepeatedHeader, null),
			m(RepeatedMain, null)
		)
	}
}

export const repeatedTree = () => m(RepeatedRoot)
